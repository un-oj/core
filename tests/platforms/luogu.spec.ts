import { NotFoundError } from '@un-oj/core';
import Luogu from '@un-oj/core/platforms/luogu';
import { describe, expect, it } from 'bun:test';

describe('Luogu platform', () => {
  const luogu = new Luogu();

  it('should fetch problem B2001', async () => {
    expect(await luogu.getProblem('B2001')).toMatchSnapshot();
  });

  it('should fetch problem P2573', async () => {
    expect(await luogu.getProblem('P2573')).toMatchSnapshot();
  });

  it('should throw NotFoundError for non-existent problem P114514', async () => {
    expect(luogu.getProblem('P114514')).rejects.toThrow(NotFoundError);
  });
});
