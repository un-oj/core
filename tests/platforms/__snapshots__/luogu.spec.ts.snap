// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Luogu platform should fetch problem B2001 1`] = `
{
  "description": {
    "background": "强烈推荐[新用户必读贴](/discuss/show/241461)",
    "details": "求两个整数的和。",
    "hint": "对于 $100\\%$ 的数据，输入的整数在 $[1, 4 \\times {10}^{18}]$ 内。",
    "input": "一行，两个用空格隔开的整数。",
    "output": "两个整数的和。",
  },
  "difficulty": 1,
  "id": "B2001",
  "link": "https://www.luogu.com.cn/problem/B2001",
  "memoryLimit": [
    134217728,
    134217728,
    134217728,
    134217728,
    134217728,
    134217728,
    134217728,
    134217728,
    134217728,
    134217728,
  ],
  "samples": [
    {
      "input": "1 2",
      "output": "3",
    },
    {
      "input": "10230 21312",
      "output": "31542",
    },
  ],
  "tags": [
    "1",
    "353",
  ],
  "timeLimit": [
    1000,
    1000,
    1000,
    1000,
    1000,
    1000,
    1000,
    1000,
    1000,
    1000,
  ],
  "title": "入门测试题目",
  "type": "traditional",
}
`;

exports[`Luogu platform should fetch problem P2573 1`] = `
{
  "description": {
    "background": "",
    "details": 
"a180285 非常喜欢滑雪。他来到一座雪山，这里分布着 $m$ 条供滑行的轨道和 $n$ 个轨道之间的交点（同时也是景点），而且每个景点都有一编号 $i\\space (1 \\le i \\le n)$ 和一高度 $h_i$。  

a180285 能从景点 $i$ 滑到景点 $j$ 当且仅当存在一条 $i$ 和 $j$ 之间的边，且 $i$ 的高度**不小于** $j$。与其他滑雪爱好者不同，a180285 喜欢用最短的滑行路径去访问尽量多的景点。如果仅仅访问一条路径上的景点，他会觉得数量太少。

于是 a18028 5拿出了他随身携带的时间胶囊。这是一种很神奇的药物，吃下之后可以立即回到上个经过的景点（不用移动也不被认为是 a180285 滑行的距离）。  

请注意，这种神奇的药物是可以连续食用的，即能够回到较长时间之前到过的景点（比如上上个经过的景点和上上上个经过的景点）。 现在，a180285站在 $1$ 号景点望着山下的目标，心潮澎湃。他十分想知道在不考虑时间胶囊消耗的情况下，以最短滑行距离滑到尽量多的景点的方案（即满足经过景点数最大的前提下使得滑行总距离最小）。你能帮他求出最短距离和景点数吗？
"
,
    "hint": 
"【数据范围】   
对于 $ 30\\% $ 的数据，$ 1 \\le n \\le 2000 $；   
对于 $ 100\\% $ 的数据，$ 1 \\le n \\le 10^5 $。


对于所有的数据，保证 $ 1 \\le m \\le 10^6 $ , $ 1 \\le h_i \\le 10^9 $ ，$ 1 \\le k_i \\le 10^9 $。
"
,
    "input": 
"输入的第一行是两个整数 $n,m$。
接下来一行有 $n$ 个整数 $h_i$，分别表示每个景点的高度。

接下来 $m$ 行，表示各个景点之间轨道分布的情况。每行三个整数 $u,v,k$，表示编号为 $u$ 的景点和编号为 $v$ 的景点之间有一条长度为 $k$ 的轨道。
"
,
    "output": 
"输出一行，表示 a180285 最多能到达多少个景点，以及此时最短的滑行距离总和。
"
,
  },
  "difficulty": 5,
  "id": "P2573",
  "link": "https://www.luogu.com.cn/problem/P2573",
  "memoryLimit": [
    131072000,
    131072000,
    131072000,
    131072000,
    131072000,
  ],
  "samples": [
    {
      "input": 
"3 3 
3 2 1 
1 2 1 
2 3 1 
1 3 10 "
,
      "output": "3 2",
    },
  ],
  "tags": [
    "6",
    "29",
    "39",
    "48",
    "113",
    "166",
  ],
  "timeLimit": [
    5000,
    5000,
    5000,
    5000,
    5000,
  ],
  "title": "[SCOI2012] 滑雪",
  "type": "traditional",
}
`;
